src/shared/discover-image.c

```diff
@@ -754,29 +753,44 @@ int image_find(RuntimeScope scope,
         if (!names)
                 return -ENOMEM;
 
+        _cleanup_close_ int root_fd = AT_FDCWD;
+        if (root) {
+                root_fd = open(root, O_CLOEXEC|O_DIRECTORY|O_PATH);
+                if (root_fd < 0) {
+                        log_debug_errno(errno, "Failed to open root directory '%s': %m", root);
+                        return -errno;
+                }
+        }
+
         _cleanup_strv_free_ char **search = NULL;
         r = pick_image_search_path(scope, class, root, &search);
         if (r < 0)
                 return r;
 
         STRV_FOREACH(s, search) {
-                _cleanup_free_ char *resolved = NULL;
                 _cleanup_closedir_ DIR *d = NULL;
 
-                r = chase_and_opendir(*s, root, CHASE_PREFIX_ROOT, &resolved, &d);
+                r = chase_and_opendir(*s, root, CHASE_PREFIX_ROOT, /* ret_path= */ NULL, &d);
                 if (r == -ENOENT)
                         continue;
                 if (r < 0)
                         return r;
 
                 STRV_FOREACH(n, names) {
-                        _cleanup_free_ char *fname_buf = NULL;
                         const char *fname = *n;
+                        _cleanup_free_ char *fname_path = NULL;
+                        _cleanup_free_ char *resolved_file = NULL;
+                        _cleanup_close_ int fd = -EBADF;
 
-                        _cleanup_close_ int fd = openat(dirfd(d), fname, O_PATH|O_CLOEXEC|open_flags);
+                        fname_path = path_join(*s, fname);
+                        if (!fname_path)
+                                return -ENOMEM;
+
+                        /* Follow symlinks only inside given root */
+                        fd = chase_and_open(fname_path, root, CHASE_PREFIX_ROOT, O_PATH|O_CLOEXEC, &resolved_file);
```

Comment:
You already opened a file descriptor to `root`, so use `chaseat` here instead of `chase`. Also you only need an `O_PATH` descriptor so there's no need for `chase_and_open()`, `chaseat()` will suffice.

Reply:
I had looked into it and think it wasn't possible due to `CHASE_PREFIX_ROOT` being needed, will try to see if it somehow works and add a comment if not